src/main/java/org/elasticsearch/common/network/MulticastChannel.java

```diff
@@ -281,12 +282,23 @@ protected void close(Listener listener) {
         }
 
         private MulticastSocket buildMulticastSocket(Config config) throws Exception {
+            SocketAddress addr = new InetSocketAddress(InetAddress.getByName(config.group), config.port);
             MulticastSocket multicastSocket = new MulticastSocket(config.port);
             try {
                 multicastSocket.setTimeToLive(config.ttl);
-                // set the send interface
-                multicastSocket.setInterface(config.multicastInterface);
-                multicastSocket.joinGroup(InetAddress.getByName(config.group));
+                // OSX is not smart enough to tell that a socket bound to the
+                // 'lo0' interface needs to make sure to send the UDP packet
+                // out of the lo0 interface, so we need to do some special
+                // workarounds to fix it.
+                if (Constants.MAC_OS_X) {
+                    // 'null' here tells the socket to deter to the interface set
+                    // with .setInterface
+                    multicastSocket.joinGroup(addr, null);
+                    multicastSocket.setInterface(config.multicastInterface);
```

Comment:
Hmm I see. I wonder if instead of having this hardcoded we can handle it with a setting that will default to the right value on Mac OS X (just in case this bug is fixed sometime in the future). Basically, can we do something similar to what we do with shared socket [here](https://github.com/elastic/elasticsearch/blob/master/src/main/java/org/elasticsearch/discovery/zen/ping/multicast/MulticastZenPing.java#L128).